File: mobile/modules/expo-squircle-view/android/src/main/java/expo/modules/squircleview/ExpoSquircleView.kt
package expo.modules.squircleview

import android.content.Context
import android.graphics.Canvas
import android.graphics.DashPathEffect
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.RectF
import android.util.DisplayMetrics
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView

/**
 * A squircle drawn straight onto the canvas. The view normally carries no
 * content of its own — it is stretched behind whatever it is meant to be the
 * background of.
 *
 * With `clipsContent` it becomes a container instead: children lay out inside it
 * and are masked to the shape, which is the only way to give a corner that has
 * to clip — a scroller, an image, a row of ticks — the same curve as its
 * neighbours.
 *
 * The stroke is inset by half its width, the way a CSS border sits inside the
 * box, so a bordered squircle occupies exactly the frame it was given.
 *
 * Props arrive in dp, as every React Native length does; the path is in pixels.
 */
class ExpoSquircleView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
    private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
    private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE }

    /**
     * Punches the corners out of the children once they are drawn. A plain
     * `clipPath` would do the same job in one step, but the hardware canvas
     * clips without antialiasing and a jagged squircle is worse than a round one.
     */
    private val erasePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
    }

    /** The shape at its full size — what the fill paints and the mask keeps. */
    private val outline = Path()

    /** The same, inset by half the stroke, so the hairline lands inside the box. */
    private val hairline = Path()

    /** Everything the mask throws away: the box minus the shape. */
    private val surround = Path()

    private var radii = SquircleRadii()
    private var smoothing = 0f
    private var clipsContent = false
    private var stale = true

    init {
        fillPaint.color = 0x00000000
        strokePaint.color = 0x00000000
        setWillNotDraw(false)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // Rebuilt here rather than in each setter: a render hands the view its
        // props one at a time, and only the last of them has the whole shape.
        if (stale) rebuild()

        canvas.drawPath(outline, fillPaint)

        // With nothing drawn over it, the hairline can go down beside the fill
        // and `dispatchDraw` is left alone.
        if (!clipsContent) drawStroke(canvas)
    }

    override fun dispatchDraw(canvas: Canvas) {
        if (!clipsContent) {
            super.dispatchDraw(canvas)
            return
        }

        if (stale) rebuild()

        val layer = canvas.saveLayer(0f, 0f, width.toFloat(), height.toFloat(), null)
        super.dispatchDraw(canvas)
        canvas.drawPath(surround, erasePaint)
        canvas.restoreToCount(layer)

        // Over the children rather than under them, or a full-bleed child would
        // paint the border out.
        drawStroke(canvas)
    }

    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
        super.onSizeChanged(width, height, oldWidth, oldHeight)
        stale = true
    }

    private fun drawStroke(canvas: Canvas) {
        if (strokePaint.strokeWidth > 0f) canvas.drawPath(hairline, strokePaint)
    }

    private fun rebuild() {
        stale = false
        outline.reset()
        hairline.reset()
        surround.reset()
        if (width == 0 || height == 0) return

        val size = { inset: Float ->
            SquirclePath.create(
                width = width - inset * 2f,
                height = height - inset * 2f,
                radii = if (inset > 0f) radii.inset(inset) else radii,
                smoothing = smoothing,
            )
        }

        size(0f).transform(Matrix(), outline)

        val inset = strokePaint.strokeWidth / 2f
        if (inset > 0f) {
            size(inset).transform(Matrix().apply { setTranslate(inset, inset) }, hairline)
        } else {
            hairline.set(outline)
        }

        surround.addRect(RectF(0f, 0f, width.toFloat(), height.toFloat()), Path.Direction.CW)
        surround.op(outline, Path.Op.DIFFERENCE)
    }

    private fun invalidateShape() {
        stale = true
        invalidate()
    }

    private fun toPixels(dp: Float): Float =
        dp * (context.resources.displayMetrics.densityDpi.toFloat() / DisplayMetrics.DENSITY_DEFAULT)

    fun setFillColor(color: Int) {
        fillPaint.color = color
        invalidate()
    }

    fun setStrokeColor(color: Int) {
        strokePaint.color = color
        invalidate()
    }

    /** Lengths of the drawn and undrawn runs, or nothing at all for a solid line. */
    fun setStrokeDash(pattern: List<Float>) {
        strokePaint.pathEffect =
            if (pattern.isEmpty()) null
            else DashPathEffect(pattern.map { toPixels(it) }.toFloatArray(), 0f)
        invalidate()
    }

    fun setStrokeWidth(dp: Float) {
        strokePaint.strokeWidth = toPixels(dp)
        invalidateShape()
    }

    fun setSmoothing(value: Float) {
        smoothing = value
        invalidateShape()
    }

    fun setClipsContent(value: Boolean) {
        clipsContent = value
        invalidate()
    }

    fun setRadii(transform: SquircleRadii.() -> SquircleRadii) {
        radii = radii.transform()
        invalidateShape()
    }

    fun setRadiusTopLeft(dp: Float) = setRadii { copy(topLeft = toPixels(dp)) }

    fun setRadiusTopRight(dp: Float) = setRadii { copy(topRight = toPixels(dp)) }

    fun setRadiusBottomRight(dp: Float) = setRadii { copy(bottomRight = toPixels(dp)) }

    fun setRadiusBottomLeft(dp: Float) = setRadii { copy(bottomLeft = toPixels(dp)) }
}

File: mobile/modules/expo-squircle-view/ios/ExpoSquircleView.swift
import ExpoModulesCore
import UIKit

/**
 A squircle painted on shape layers. The view normally carries no content of its
 own — it is stretched behind whatever it is meant to be the background of, so
 the fill and the stroke land on native layers rather than on a measured SVG.

 With `clipsContent` it becomes a container instead: children lay out inside it
 and are masked to the shape, which is the only way to give a corner that has to
 clip — a scroller, an image, a row of ticks — the same curve as its neighbours.

 The stroke is inset by half its width, the way a CSS border sits inside the box,
 so a bordered squircle occupies exactly the frame it was given.
 */
class ExpoSquircleView: ExpoView {
  private let fill = CAShapeLayer()
  private let stroke = CAShapeLayer()
  private let clip = CAShapeLayer()

  var radii = SquircleRadii() { didSet { setNeedsLayout() } }
  var cornerSmoothing: CGFloat = 0 { didSet { setNeedsLayout() } }
  var clipsContent = false { didSet { setNeedsLayout() } }

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    for shape in [fill, stroke] {
      shape.fillColor = UIColor.clear.cgColor
      shape.strokeColor = UIColor.clear.cgColor
      layer.addSublayer(shape)
    }
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override func layoutSubviews() {
    super.layoutSubviews()

    // Implicit animations would trail the shape behind a resize by a quarter
    // second, which reads as the background lagging its own view.
    withoutAnimation {
      let outline = path(inset: 0)

      fill.frame = bounds
      fill.path = outline

      stroke.frame = bounds
      stroke.path = stroke.lineWidth > 0 ? path(inset: stroke.lineWidth / 2) : outline

      clip.frame = bounds
      clip.path = outline
      layer.mask = clipsContent ? clip : nil
    }

    raiseStroke()
  }

  /// A child arrives as a subview, so its layer lands above the ones added here.
  override func didAddSubview(_ subview: UIView) {
    super.didAddSubview(subview)
    raiseStroke()
  }

  /// The hairline belongs over whatever the view is clipping, not under it.
  private func raiseStroke() {
    guard clipsContent, layer.sublayers?.last !== stroke else { return }
    withoutAnimation { layer.addSublayer(stroke) }
  }

  private func path(inset: CGFloat) -> CGPath {
    let shape = SquirclePath.create(
      width: bounds.width - inset * 2,
      height: bounds.height - inset * 2,
      radii: inset > 0 ? radii.inset(by: inset) : radii,
      smoothing: cornerSmoothing
    )

    guard inset > 0 else { return shape }

    var shift = CGAffineTransform(translationX: inset, y: inset)
    return shape.copy(using: &shift) ?? shape
  }

  func setFillColor(_ color: UIColor) {
    withoutAnimation { fill.fillColor = color.cgColor }
  }

  func setStrokeColor(_ color: UIColor) {
    withoutAnimation { stroke.strokeColor = color.cgColor }
  }

  /// Lengths of the drawn and undrawn runs, or nothing at all for a solid line.
  func setStrokeDash(_ pattern: [CGFloat]) {
    withoutAnimation {
      stroke.lineDashPattern = pattern.isEmpty ? nil : pattern.map { NSNumber(value: Double($0)) }
    }
  }

  func setStrokeWidth(_ width: CGFloat) {
    withoutAnimation { stroke.lineWidth = width }
    setNeedsLayout()
  }

  private func withoutAnimation(_ body: () -> Void) {
    CATransaction.begin()
    CATransaction.setDisableActions(true)
    body()
    CATransaction.commit()
  }
}

private extension SquircleRadii {
  /// The centreline of a stroke runs half a width inside each corner's arc.
  func inset(by amount: CGFloat) -> SquircleRadii {
    SquircleRadii(
      topLeft: max(0, topLeft - amount),
      topRight: max(0, topRight - amount),
      bottomRight: max(0, bottomRight - amount),
      bottomLeft: max(0, bottomLeft - amount)
    )
  }
}
